File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.py
def solution(cacheSize, cities):
    answer = 0
    cache = []

    for city in cities:
        city = city.lower()
        if city in cache:
            answer += 1
            cache.remove(city)
            cache.append(city)
        else:
            answer += 5
            if len(cache) < cacheSize:
                cache.append(city)
            elif cacheSize != 0:
                cache.pop(0)
                cache.append(city)

    return answer

File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int solution(int cacheSize, String[] cities) {
        if (cacheSize == 0) {
            return cities.length * 5;
        }

        Queue<String> cache = new LinkedList<>();
        int answer = 0;

        for (String city : cities) {
            String cityLower = city.toLowerCase();
            
            // cache hit
            if (cache.contains(cityLower)) {
                cache.remove(cityLower);
                cache.offer(cityLower);
                answer += 1;
            }
            // cache miss
            else {
                if (cache.size() == cacheSize) {
                    cache.poll();
                }
                cache.offer(cityLower);
                answer += 5;
            }
        }

        return answer;
    }
}
